Improve comparison of morphisms of modules

HechtiDerLachs opened the pull request:
This is to address one of the open points in #1806.

HechtiDerLachs committed:
Fix up comparison of maps of modules.

HechtiDerLachs committed:
Add tests.

HechtiDerLachs commented on src/Modules/UngradedModules/SubQuoHom.jl:
```diff
@@ -329,14 +329,33 @@ function is_welldefined(H::SubQuoHom)
+# TODO: Move to Hecke?
+function (==)(f::Map, g::Map)
+  f === g && return true
+  error("comparison of maps of type $(typeof(f)) and $(typeof(g)) not implemented")
```
@thofma : Is this legitimate? Or did you decide other things in Hecke?

thofma replied:
Hm, I am not really a fan of this catch all fallback. It just hides problems and effects everyone. If a map type `MyMap` does not want to implement `==` properly, it should do this thing just for `MyMap`.

thofma replied:
Happy to hear what others think @fingolfin @lgoettgens maybe?

lgoettgens replied:
really not a fan of this. and in the small chance that we want to have it, it should definitely put into AA (as this is one of the worst instances of type piracy I can think of)

HechtiDerLachs replied:
Ok, thanks for the feedback. It felt weird to implement this method of `==` (and I'm aware of the type piracy, see the preceding TODO), but it nevertheless seemed like the first plausible attempt to a solution. 

The problem I see here is the following. We deliberately allow *everything* as coefficient maps, both for maps of polynomial rings and friends, but also for morphisms of modules. So --as it at least seems to me-- to "fix its `==`-problems" (see @thofma 's comment below) one would need to implement `(==)(R::MPolyRing, a::Any)`, just to catch comparison of `f` and `g` where `f` has an `MPolyRing` used as a type-caster as an internal map and `g` maybe has an anonymous function. Now what should the above method return? 

One could say `false`, because an `MPolyRing` is equal to another object if and only if they are the same instance in memory and another method `(==)(::MPolyRing, ::MPolyRing)` takes care of this. 

One could say, it should error, because there is no way to check whether `R` seen as a map is mathematically equal to `a`. 

But then what about `(==)(::MPolyRing, ::Map)`? A ring should probably never be equal to a map, but in this context it seems to be required by "fixing its `(==)`-problems". Since the `Map` has the information about its 
domain, the effect of the `MPolyRing` and the `Map` can probably be compared and mathematical equality of their effect can be decided in many instances. Shouldn't we overwrite `==` then? 

I don't claim that I have any solution and I don't draw pleasure from overthrowing your remarks. But it seems to me that things are a bit tricky here. And I do not see straight away how to resolve this.

thofma replied:
OK, I see some of the problems, but how does defining `==(x::Map, y::Map)` solve anything related to `==(x::Map, ::MPolyRing)`?

How about checking `===` of the coefficient maps directly in https://github.com/oscar-system/Oscar.jl/blob/master/src/Rings/mpolyquo-localizations.jl#L2693 or where needed?

Edit: Or how about adding
```
isegal_or_error(x, y) == x === y ? true : error("cannot be compared")
```
and use this where needed?

HechtiDerLachs replied:
> how does defining `==(x::Map, y::Map)` solve anything related to `==(x::Map, ::MPolyRing)`?

It doesn't. When I wrote that method, I had not yet fully explored the abyss, but just thought to take an optimistic step forward. Futile, as it seems. 

Yes, doing things like `===` comparison internally seems reasonable. But it feels like we should pin something down as to what kind of comparisons we want to allow and which ones not. For instance, we could allow comparison only for maps of modules `f` and `g` which both have a `Map` for their rings and then require actual `Map`s to solve their `==`-problems. Then we would throw a not-implemented-error from within the implementation for `f==g` and not delegate this further. Does that seem reasonable?  

Edit: I just see it now (again?): what you wrote below seems to take a similar direction, right? I.e. restricting the type for admissible comparison.

thofma replied:
Yes, this sounds reasonable.

To be honest, I kind of regret allowing everything for the coefficient rings. But it might be a bit late to go back.

HechtiDerLachs commented on src/Modules/UngradedModules/SubQuoHom.jl:
```diff
@@ -329,14 +329,33 @@ function is_welldefined(H::SubQuoHom)
+    isnothing(ring_map(f)) || is_trivial(ring_map(f)) || return false
+  end
+  # TODO: Catch the case where == defaults to === and this returns false, 
+  # but the ring_maps are identical in the mathematical sense nevertheless.
```
I am a bit puzzled with the problem we're facing here. I would like to throw an error in case comparison of maps is not implemented. But given the freedom that we gave to the maps of rings, it seems unavoidable that for numerous combinations `==` will just default to `===` and then return `false`, even though the maps might be mathematically the same or not comparable. 

But this must have popped up elsewhere, too. How is it solved, for instance, for comparison of morphisms of polynomial rings?

thofma replied:
<s>Comparison of the images and then just `==` for the maps of the coefficient rings. Either the coefficient ring map is `nothing` (aka identity), or it is some proper type, in which case we expect it to behave nice. If it does not, the coefficient ring map needs to fix its `==` problems.</s>

Not quite right. I guess you implemented this here https://github.com/oscar-system/Oscar.jl/blob/master/src/Rings/mpolyquo-localizations.jl#L2693 for polynomial rings.

HechtiDerLachs replied:
> I guess you implemented this here

Indeed, I did. And I didn't do it properly and had to correct it here in the latest version.

HechtiDerLachs committed:
Remove deprecated code.

HechtiDerLachs committed:
Update after discussion.

HechtiDerLachs committed:
Adapt comparison of some ring maps.

HechtiDerLachs committed:
Update tests.

HechtiDerLachs commented:
I made an attempt to switch to allowing comparison only in cases where the `ring_map`s are either `nothing` or an honest `Map`. In the other cases, I force `===` or throw an error, informing the user that we have no means to compare such maps in general. 

But even restricting to `Map`s requires some things to be fulfilled. For instance, I didn't find `MapFromFunc` to work for this out of the box and I had to overwrite `==` for them. So it seems to me we might need to formulate the interface/behavior of `Map`s a little more rigorous. 

Anyway: Let me know what you think about the current proposal.

lgoettgens commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2699,27 +2699,50 @@ function Base.:(==)(
+  codomain(f) === codomain(g) || return false
+  f.f === g.f && return true
+  error("comparison of maps $f and $g not possible")
+end
```
This does not look like a great solution for me, since it most cases the `.f` fields will just be anonymous functions that are not comparable.
As such, I would find it more natural for MapFromFunc to have no custom `==` at all, thus falling back to `===`

HechtiDerLachs replied:
> As such, I would find it more natural for MapFromFunc to have no custom `==` at all, thus falling back to `===`

We could do that, yes. But in that case it is not possible to require `==` to have a mathematically correct implementation for `Map`s in general. I.e. we can not restrict to cases with some type-parameter being `<:Map` and then safely assume that `==` can be used in the body of the function. 

To me `Map` seemed like a good cutting point to distinguish cases which can be handled from those which can't. But with your suggestion, I'm afraid this would fall. 

Edit: Let me add that my implementation here is probably the only *mathematically correct* implementation here, while defaulting to `===` is probably not.

lgoettgens replied:
I think we had this discussion already about 2 years ago, but unfortunately I cannot find it anymore.
What I think we concluded to (but this could also be totally wrong):
`==` can in many cases not be any mathematical equivalence. Because what should e.g. `G == H` for two groups do? Check if they are isomorphic? That's way more complicated than what we usually want `==` to be.
Thus, we use `==` more as a object equality, i.e. the two objects are somehow defined in the same way and thus seem to be equivalent. This is, in general, nothing *mathematically sane* to reason about.

But I think this here is such an occurrence where we just *cannot* do something mathematically sane, since anonymous function can do too many things.

thofma replied:
@lgoettgens is your point that maps should be compared like groups? I don't remember that we agreed on this.

If map/homomorphism comparison should be doing something mathematical, then the MapFromFunc should compare equal if they are egal (`===`) and it should throw an error otherwise. (which is what @HechtiDerLachs is advocating for here, and with which I agree.)

HechtiDerLachs replied:
Ok. It would be good if we could dig out the conclusions of that discussion again. I agree that there are cases where we can not reasonably check mathematical equality with generic code. Hence, the mathematically rigorous solution would be to throw an error in these cases (I think). 

In general, one can, of course, communicate to the user that e.g. for groups `==` is not implemented in the mathematical way. But in this case here (morphisms of modules), one would probably expect the mathematically rigorous answer. 

Another possible solution that I see is to not allow `==` for morphisms of modules with a `ring_map` at all (and probably similar for morphisms of `MPolyAnyRing`s?). But then again, I would not write it somewhere in the docstring that "the user should be aware that mathematical comparison via `==` only works in abscence of a `ring_map`". Since that distinction is not tied to the outer type, but only to one of its parameters, the user will probably not find that information. So even there I would implement `==` to throw an error.

lgoettgens replied:
> the MapFromFunc should compare equal if they are egal (`===`) and it should throw an error otherwise.

I don't object adding this to Hecke.

I am not even sure myself what point I wanted to make above... hmm...

thofma replied:
Just to make sure. Is something failing here without this `==` for `MapFromFunc`? I might be tempted to make it to always return `false` otherwise (until someone has an application).

HechtiDerLachs replied:
Yes: The tests as they are now would fail without this implementation. In general, my code is using `MapFromFunc` in many places as maps for coefficients. Not sure how much comparison is done with these objects, in the end, though. 

If there are good reasons for always returning `false`, then I would have to revise this. But otherwise, I would prefer to keep it like this.

fingolfin commented:
Summarizing an example by @HechtiDerLachs during triage.

Cosider $\phi,\psi: \mathbb{Z}^2 \to \mathbb{Q}^3$.

$\phi(e_1)=e_1, \phi(e_2)=e_2$.

$\psi(e_1)=e_1, \psi(e_2)=e_3$.

Should we be able to compare $\phi$ and $\psi$ to determine they are not equal?

HechtiDerLachs committed:
Use is_equal_as_morphism [no ci].

lgoettgens commented on src/Modules/UngradedModules/SubQuoHom.jl:
```diff
@@ -329,36 +329,98 @@ function is_welldefined(H::SubQuoHom)
+function (==)(f::ModuleFPHom, g::ModuleFPHom)
+  # `===` is the only generic case we can handle in general; see above comment.
+  is_equal_as_morphism(f, g) || return false # this call might even throw!
   return true
```
```suggestion
  return is_equal_as_morphism(f, g) # this call might even throw!
```
that should be equivalent, but more readable

HechtiDerLachs committed:
Update code and comments.

HechtiDerLachs commented:
This now depends on https://github.com/Nemocas/AbstractAlgebra.jl/pull/2035.

joschmitt commented:
https://github.com/Nemocas/AbstractAlgebra.jl/pull/2035 has been merged and released some time ago. So this should work now?

HechtiDerLachs committed:
Reroute comparison of Maps.

HechtiDerLachs committed:
Update tests.

HechtiDerLachs commented:
I have rebased this on the latest master and updated the code and the tests. Some tests do not seem to run, yet, but I do not really understand what the issue is. @benlorenz , @aaruni96 : Can you maybe tell me more? Thx!

Edit: Once the tests pass, I will eventually look into preparing some release notes. Let me know if you think that this is indeed necessary.

lgoettgens commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2698,27 +2698,54 @@ function Base.:(==)(
+end
+
+# TODO: Move to wherever `MapFromFunc` comes from ?
+function is_equal_as_morphism(f::MapFromFunc, g::MapFromFunc)
```
you can move this function to `Misc/MoveToHecke.jl`, then we remember to move it to Hecke once we do another large moving upstream attempt.

lgoettgens commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2698,27 +2698,54 @@ function Base.:(==)(
+  h = hash(domain(f), h)
+  h = hash(codomain(f), h)
+  return h
 end
```
```suggestion
```
since there is no `==` anymore for `MapFromFunc`, this `hash` can go as well

lgoettgens commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2698,27 +2698,54 @@ function Base.:(==)(
+  domain(f) === domain(g) || return false
+  codomain(f) === codomain(g) || return false
+  error("comparison of maps $f and $g not possible")
+end
```
```suggestion
```
I think these are no longer needed here, as `is_equal_as_morphism` takes care of these cases already

HechtiDerLachs replied:
I found them to be necessary for some reason. But I will look into this again.

lgoettgens reviewed (commented):
There is a genuine test failure in https://github.com/oscar-system/Oscar.jl/actions/runs/14620509078/job/41019057548?pr=4706#step:11:1724 `comparison of maps Map: finite quadratic module -> finite quadratic module and Map: finite quadratic module -> finite quadratic module not possible`

HechtiDerLachs committed:
Make some more comparisons work.

HechtiDerLachs committed:
Adjust tests.

HechtiDerLachs commented:
> Should we be able to compare ϕ and ψ to determine they are not equal?

Now the following code runs:
```julia
R, (x, y) = ZZ[:x, :y]
S, (u, v) = QQ[:u, :v]
R1 = free_module(R, 1)
S1 = free_module(S, 1)
phi = hom(R, S, gens(S)) # Implicit casting by `QQ`
psi = hom(R, S, QQ, gens(S)) # Explicit casting by `QQ`
f = hom(R1, S1, gens(S1), phi)
g = hom(R1, S1, [zero(S1)], phi)
@test !(f == g) # comparison should work
h = hom(R1, S1, gens(S1), psi)
@test f == h # here too
@test !(h == g)
```

HechtiDerLachs commented:
> There is a genuine test failure in...

Somehow I am unable to reproduce the failure. The tests start producing trouble at other points already. Let's see what this run says, though...

Edit: When I run the test locally, I get the following error from line 300 of `experimental/QuadFormAndIsom/test/runtests.jl`:
```
julia> ok, sv = primitive_embeddings(rescale(E8, 2), rescale(k, QQ(1//2)); check=false)
ERROR: MethodError: no method matching hom(::TorQuadModule, ::TorQuadModule, ::Vector{TorQuadModuleElem}, ::Vector{TorQuadModuleElem})
The function `hom` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  hom(::TorQuadModule, ::TorQuadModule, ::Vector{TorQuadModuleElem})
   @ Hecke ~/.julia/packages/Hecke/G43QZ/src/QuadForm/Torsion.jl:683
  hom(::TorQuadModule, ::TorQuadModule, ::ZZMatrix)
   @ Hecke ~/.julia/packages/Hecke/G43QZ/src/QuadForm/Torsion.jl:670
  hom(::MPolyQuoRing, ::NCRing, ::Any, ::Vector; check)
   @ Oscar ~/Oscar.jl/src/Rings/MPolyMap/MPolyQuo.jl:78
  ...

Stacktrace:
 [1] _sum_with_embeddings_orthogonal_groups(A::TorQuadModule, B::TorQuadModule)
   @ Oscar ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:30
 [2] _glue_stabilizers(L::ZZLatWithIsom, M::ZZLatWithIsom, N::ZZLatWithIsom; subM::Bool, subN::Bool)
   @ Oscar ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:2071
 [3] _glue_stabilizers
   @ ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:2057 [inlined]
 [4] #_glue_stabilizers#9338
   @ ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:2100 [inlined]
 [5] _glue_stabilizers
   @ ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:2096 [inlined]
 [6] primitive_embeddings(G::ZZGenus, M::ZZLat; classification::Symbol)
   @ Oscar ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:1411
 [7] primitive_embeddings
   @ ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:1290 [inlined]
 [8] primitive_embeddings(L::ZZLat, M::ZZLat; classification::Symbol, check::Bool)
   @ Oscar ~/Oscar.jl/experimental/QuadFormAndIsom/src/embeddings.jl:1220
 [9] top-level scope
   @ REPL[7]:1
```
I do not see how that can be related with the changes made here. @StevellM : Any ideas?

StevellM commented:
I guess you probably need to update your Hecke version. There has been a fix following some Hecke changes, but the fix works only if Hecke is recent enough (0.35.17 at least)

HechtiDerLachs committed:
Attempt to fix failures in the lattices.

HechtiDerLachs commented:
@StevellM : Some of your tests seem to be randomized. I had to run some tests multiple times to provoke failures and I can not be sure 100% that the problem is fixed now or whether I should simply try more often to make it fail. Is there a way to un-randomize things?

StevellM commented:
Do you have concrete examples of tests failing in `QuadFormAndIsom`, different from the previous one ? As far as I know we have not experienced test failure due to randomness. We have recently fixed some things on Hecke, which could produce weird behavior, but I doubt the current tests even touched these pieces of code. The very few places where we may have randomness are barely tested, so I do not think it is relevant to "untest" them.

HechtiDerLachs commented:
> Do you have concrete examples of tests failing in `QuadFormAndIsom`

Yes, sorry for being inprecise. Try the testset `"Primitive extensions and embeddings"` starting in line 293 of your test file on the second last commit ([49b14dc](https://github.com/oscar-system/Oscar.jl/pull/4706/commits/49b14dc7179488bc3b93a642184258ce15c5a2ab)) in this PR. I got random failures there. 

Edit: It does not seem to be too urgent as tests are not failing anymore now. But there are general efforts to avoid randomness in tests, so I thought, it might be good to point that out.

HechtiDerLachs commented:
Either way: Things seem to be fine now. Waiting for approval and merge from my side.

HechtiDerLachs committed:
Move some methods to Hecke-file.

fingolfin commented:
@lgoettgens will look at this today and hopefully we can merge it later

lgoettgens committed:
Move a method

lgoettgens reviewed (approved):
no objections, I just moved a method to the right file. Enabling auto-merge

HechtiDerLachs committed:
Merge branch 'master' into fixes_for_modules
